### PR TITLE
feat(dsp_enhancement): create dsp cr for pools not managed by dsp

### PR DESCRIPTION
**Pull request by abhilashshetty04**

### CODE REVIEW

**Comment by abhilashshetty04** [comment]
bors try

File: k8s/operators/src/pool/diskpool/client.rs [hunk 1]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
     Ok(())
 }
 
+/// List pool from control plane. Checks if DSP cr present for it. If no, create it ourself.
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]: Something like this:

```suggestion
/// Reconciles control-plane pools into CRs by listing pools from the control plane and creating equivalent CRs if respective CR is not present.
```

File: k8s/operators/src/pool/diskpool/client.rs [hunk 2]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+pub(crate) async fn create_missing_cr(
+    k8s: &Client,
+    control_client: clients::tower::ApiClient,
+    namespace: &str,
+) -> Result<(), Error> {
+    let pools = control_client.pools_api().get_pools().await.unwrap();
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]: handle unwrap

File: k8s/operators/src/pool/diskpool/client.rs [hunk 3]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+    let pools_api: Api<DiskPool> = v1beta1_api(k8s, namespace);
+    let param = PostParams::default();
+    for pool in pools.into_body().iter_mut() {
+        match pools_api.get(&pool.id).await {
+            Err(kube::Error::Api(e)) if e.code == StatusCode::NOT_FOUND => {
+                info!(pool.id, "Found pool to be missing in k8s");
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]:
```suggestion
                tracing::warn!(pool.id, pool.node, "DiskPool CR is missing");
```

File: k8s/operators/src/pool/diskpool/client.rs [hunk 4]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+                if let Some(spec) = pool.spec.clone() {
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]:
```suggestion
                if let Some(spec) = pool.spec {
```

File: k8s/operators/src/pool/diskpool/client.rs [hunk 5]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+                    let spec: DiskPoolSpec = DiskPoolSpec::new(spec.node, spec.disks);
+                    let new_disk_pool: DiskPool = DiskPool::new(&pool.id, spec);
+                    if let Err(error) = pools_api.create(&param, &new_disk_pool).await {
+                        info!(
+                            pool.id,
+                            "Failed creating missing pool CR with error : {:?}", error
+                        );
+                    }
+                } else {
+                    info!(
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]: hmm should we check this at the top and just ignore these pools?

File: k8s/operators/src/pool/diskpool/client.rs [hunk 6]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+            Err(kube::Error::Api(e)) if e.code == StatusCode::NOT_FOUND => {
+                info!(pool.id, "Found pool to be missing in k8s");
+                if let Some(spec) = pool.spec.clone() {
+                    let spec: DiskPoolSpec = DiskPoolSpec::new(spec.node, spec.disks);
+                    let new_disk_pool: DiskPool = DiskPool::new(&pool.id, spec);
+                    if let Err(error) = pools_api.create(&param, &new_disk_pool).await {
+                        info!(
+                            pool.id,
+                            "Failed creating missing pool CR with error : {:?}", error
+                        );
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]:
```suggestion
                        info!(
                            pool.id,
                            pool.node,
                            error,
                            "Failed to create CR for missing DiskPool"
                        );
```

File: k8s/operators/src/pool/diskpool/client.rs [hunk 7]
```diff
@@ -178,6 +178,41 @@ pub(crate) async fn migrate_to_v1beta1(
...
+    let pools = control_client.pools_api().get_pools().await.unwrap();
+    let pools_api: Api<DiskPool> = v1beta1_api(k8s, namespace);
+    let param = PostParams::default();
+    for pool in pools.into_body().iter_mut() {
+        match pools_api.get(&pool.id).await {
+            Err(kube::Error::Api(e)) if e.code == StatusCode::NOT_FOUND => {
+                info!(pool.id, "Found pool to be missing in k8s");
+                if let Some(spec) = pool.spec.clone() {
+                    let spec: DiskPoolSpec = DiskPoolSpec::new(spec.node, spec.disks);
+                    let new_disk_pool: DiskPool = DiskPool::new(&pool.id, spec);
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]: Should we add any labels here to reflect that it was automatically created?
CC @Abhinandan-Purkait

File: k8s/operators/src/pool/diskpool/client.rs (hunk 6 above)
Comment (abhilashshetty04) [reply k8s/operators/src/pool/diskpool/client.rs]: Do we need node? That would not have any bearing on CR create fail?

File: k8s/operators/src/pool/diskpool/client.rs (hunk 7 above)
Comment (abhilashshetty04) [reply k8s/operators/src/pool/diskpool/client.rs]: @tiagolobocastro  As discussed, We will add this later when we will have operations around labels in CR.

File: k8s/operators/src/pool/diskpool/client.rs (hunk 5 above)
Comment (abhilashshetty04) [reply k8s/operators/src/pool/diskpool/client.rs]: Will be removing this arm

File: k8s/operators/src/pool/diskpool/client.rs (hunk 6 above)
Comment (abhilashshetty04) [reply k8s/operators/src/pool/diskpool/client.rs]: Added

File: k8s/operators/src/pool/diskpool/client.rs [hunk 8]
```diff
@@ -178,6 +178,36 @@ pub(crate) async fn migrate_to_v1beta1(
...
+    namespace: &str,
+) -> Result<(), Error> {
+    if let Ok(pools) = control_client.pools_api().get_pools().await {
+        let pools_api: Api<DiskPool> = v1beta1_api(k8s, namespace);
+        let param = PostParams::default();
+        for pool in pools.into_body().iter_mut() {
+            match pools_api.get(&pool.id).await {
+                Err(kube::Error::Api(e)) if e.code == StatusCode::NOT_FOUND => {
+                    if let Some(spec) = &pool.spec {
+                        warn!(pool.id, spec.node, "Found pool to be missing in k8s");
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]:
```suggestion
                        warn!(pool.id, spec.node, "DiskPool CR is missing");
```

File: k8s/operators/src/pool/diskpool/client.rs [hunk 9]
```diff
@@ -178,6 +178,36 @@ pub(crate) async fn migrate_to_v1beta1(
...
+                        let cr_spec: DiskPoolSpec =
+                            DiskPoolSpec::new(spec.node.clone(), spec.disks.clone());
+                        let new_disk_pool: DiskPool = DiskPool::new(&pool.id, cr_spec);
+                        if let Err(error) = pools_api.create(&param, &new_disk_pool).await {
+                            info!(pool.id, spec.node, %error, "Failed creating missing pool CR with error");
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/client.rs]:
```suggestion
                            error!(pool.id, spec.node, %error, "Failed to create CR for missing DiskPool");
```

**Review by tiagolobocastro — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**